Feature/sprintf lpc limits

beinvisible opened the pull request:
This implements additional formatting options for sprintf's %O and %Q modifiers.

- `%nO`/`%nQ`: 'field width' is interpreted as maximum recursion depth into arrays/mappings/structs.
- `%.nO`/`%.nQ`: 'precision' is interpreted as maximum number of array/mapping/struct elements to print.
- `%n.nO`/`%n.nQ`: both limits can be set independently, or.
- `%:nO`/`%:nQ`: both limits can be set simultaneously, as usual.

If the one of the limits is reached, further elements/substructures are omited and "..." is printed instead.

beinvisible commented:
> This changes the behavior of something like "%-=70Q". Even though that's not documented it works that way since it was implemented in LPMud 3 and it's the same in MudOs/FluffOs.

not only is it not documented, the documentation explicitly says

` "="   column mode.  Ignored unless the argument type specifier is 's'. `

> I also looked at our mudlib and it's used there. So I'm hesitant to change that.

Hmm. Well, I could revert back to the original behavior if column mode is selected.

beinvisible commented:
Ah, I see. The line 
```
                    finfo &= ~INFO_TABLE; /* since we fall through */
```

is from the master branch. I.e. the "#" modifier **is** in fact ignored, we're only taking about field width in general.
Well, yes, that would in fact be incompatible.

amotzkau commented:
But even ignoring the column mode, something like sprintf("%.50Q...", value) is in the realm of useful applications where this would change the intuitive and long standing behavior.

beinvisible commented:
How about a modifier to explicitly change the meaning of digits?

e.g.
```
sprintf("%50&3.5Q", var); // 50 character width, max. 3 levels deep, max. 5 elements 
```

That shouldn't break anything.

beinvisible commented:
implemented new modifier '&n.n' for recursion depth and/or element limit. Also cleaned up the part where the numbers are parsed in the process, which had some duplicate code.

amotzkau commented:
The format string looks awfully complicated, especially when your starting to mix those 4 options ((column) width, truncation length, element limit and recursion depth). I'm worried about its readability.

That's why I would like to make a different suggestion. For such optional, rarely used arguments in LDMud 3.6.7 we are starting to use structs as efun arguments. For sprintf() this could mean to have a modifier that just says, read additional formatting options from the next argument, and that argument needs to be a certain struct (similar to '*' which reads an width/precision from the next argument).

So it would look something like `sprintf("%&Q", (<sprintf_formating_options> recusion_depth = 2), val)`. Admittedly, this is a bit more verbose to write, but I'm expecting this not to be a very frequently used feature, and in those places it's much better to read. And it's extensible.

beinvisible commented:
Actually the struct could be used for _all_ format options, also internally (i.e. the parser would just update values in it). This would potentially clean up the code even more.

Maybe just call it (<printf_fmt>), include everything that currently flying around in separate variables, and provide defines for all fixed options. As an extreme example:

```sprintf("%&", (<printf_fmt> type: FMT_LPC, fieldsize: 10, align: FMT_CENTER, recursion: 3), val);```

amotzkau commented:
I concur with adding the existing format options there as well.  I just have reservations about the shorter struct name, but the LDMud dev team is not unanimous about that.

beinvisible commented:
I'm closing this, as the suggested solution is completely different from the code in the branch.
